File: apps/rules-service-python/src/rules.py
from typing import Optional

from pipeline.v1 import pipeline_pb2

MIN_VALUE = 10


def apply_rules(event: pipeline_pb2.ParsedEvent) -> Optional[pipeline_pb2.EnrichedEvent]:
    value = int(event.value)
    passed = value >= MIN_VALUE and event.type != "view"

    if not passed:
        return None

    return pipeline_pb2.EnrichedEvent(
        event=event,
        metadata={"rule": "min_value_and_type"},
        passed_rules=True,
    )

File: apps/demo-domain/rules-service/src/rules.ts
import type {
  EnrichedEvent,
  ParsedEvent,
} from '../../../../packages/proto/generated/ts/pipeline/v1/pipeline'

const MIN_VALUE = 10

const parseValue = (value: string): number => {
  const parsed = Number(value)
  if (!Number.isFinite(parsed)) {
    return 0
  }
  return Math.trunc(parsed)
}

/**
 * Applies simple demo rules to a parsed event.
 * @param event Parsed event payload.
 * @returns EnrichedEvent when rules pass; otherwise null.
 */
export const applyRules = (event: ParsedEvent): EnrichedEvent | null => {
  const numericValue = parseValue(event.value)
  const passed = numericValue >= MIN_VALUE && event.type !== 'view'

  if (!passed) {
    return null
  }

  return {
    event,
    metadata: {
      rule: 'min_value_and_type',
    },
    passedRules: true,
  }
}
